certora/specs/AccessManaged.spec

```
@@ -32,3 +36,24 @@ rule callRestrictedFunction(env e) {
         )
     );
 }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: Only valid authorities can be set by the current authority                                                    │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule setAuthority(env e) {
+    require nonpayable(e);
+
+    address newAuthority;
+
+    address previousAuthority = authority();
+
+    setAuthority@withrevert(e, newAuthority);
+    bool success = !lastReverted;
+
+    assert (success && authority() == newAuthority) <=> (
+        previousAuthority == e.msg.sender &&
+        _hasCode(newAuthority)
+    );
```

I would do it like this:

```suggestion
    assert success <=> (
        previousAuthority == e.msg.sender &&
        _hasCode(newAuthority)
    );
    assert success => newAuthority == authority();
```

Reason: "success && authority() == newAuthority" implies that there is a possibility that both values are different. Like the call did not revert, but the authority was not set...

---

Right, it's not possible to make a successful call and having the authority not set. I was thinking the prover will fail for the case when the new authority is the same as the previous one.

I'm fine with this one.